py/ringbuf.c

```diff
@@ -118,3 +118,114 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
     r->iput = iput_a;
     return 0;
 }
+
+#if MICROPY_PY_MICROPYTHON_RINGBUFFER
+#include "py/runtime.h"
+#include "py/stream.h"
+#include "py/mphal.h"
+
+typedef struct _micropython_ringbuffer_obj_t {
+    mp_obj_base_t base;
+    ringbuf_t ringbuffer;
+} micropython_ringbuffer_obj_t;
+
+static mp_obj_t micropython_ringbuffer_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    mp_int_t buff_size = -1;
+    mp_buffer_info_t bufinfo = {NULL, 0, 0};
+
+    if (!mp_get_buffer(args[0], &bufinfo, MP_BUFFER_RW)) {
+        buff_size = mp_obj_get_int(args[0]);
+    }
+    micropython_ringbuffer_obj_t *self = mp_obj_malloc(micropython_ringbuffer_obj_t, type);
+    if (bufinfo.buf != NULL) {
+        // buffer passed in, use it directly for ringbuffer.
+        // This can be user to no-copy stream an existing data buffer (except final byte).
+        self->ringbuffer.buf = bufinfo.buf;
+        self->ringbuffer.size = bufinfo.len;
+        self->ringbuffer.iget = self->ringbuffer.iput = 0;
+    } else {
+        // Allocate new buffer, add one extra to buff_size as ringbuf consumes one byte for tracking.
+        ringbuf_alloc(&(self->ringbuffer), buff_size + 1);
+    }
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static mp_uint_t micropython_ringbuffer_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
+    micropython_ringbuffer_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    size = MIN(size, ringbuf_avail(&self->ringbuffer));
+    uint8_t *dest = buf_in;
+    for (mp_uint_t i = 0; i < size; i++) {
+        *dest++ = ringbuf_get(&(self->ringbuffer));
+    }
```

Comment:
@dpgeorge I realised a good chunk of the extra size in this PR was due to me "reusing" the `ringbuf_get_bytes()` and `ringbuf_put_bytes()` [functions above](https://github.com/micropython/micropython/blob/a20fca8c04f95fbf9420d3c6926fd59aa6086c83/py/ringbuf.c#L82)

I hadn't realised these aren't currently used anywhere else other than in espnow, so are stripped out of most builds currently.

Replacing their use with a simple byte by byte loop is probably slightly slower, but considerably smaller.

Reply:
@andrewleech Ring buffer has many use cases in our projects in the core components. Making these simple functions slower will effect not only esp now but other use cases too.
Beside that, In many chips increasing firmware size doesn't matter that much. Maybe we can choose between the slow and fast version with a flag?

Reply:
I haven't touched the function used by espnow.

I might need to think about how to profile the performance, because I'm not confident the current byte by byte method would be much slower than the memcpy one earlier, especially with the extra "stuff" going on around the memcpy version above. 

Currently all stdio goes via the same byte by byte ringbuffer approach and it doesn't seem too bad ...